File: problems/problems_3085/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

using namespace std;
using json = nlohmann::json;

class Solution {
public:
  int minimumDeletions(string word, int k) {
    array<int, 26> freq = {0};
    for (char c : word) {
      freq[c - 'a']++;
    }
    sort(freq.begin(), freq.end());
    int ans = word.length();
    for (int i = 0; i < 26; ++i) {
      int cur = 0, maxFreq = freq[i] + k;
      for (int j = 0; j < i; ++j) {
        if (freq[j] >= freq[i]) {
          break;
        }
        cur += freq[j];
      }
      for (int j = 25; j > i; --j) {
        if (freq[j] <= maxFreq) {
          break;
        }
        cur += freq[j];
      }
      ans = min(ans, cur);
    }
    return ans;
  }
};

json leetcode::qubh::Solve(string input_json_values) {
  vector<string> inputArray;
  size_t pos = input_json_values.find('\n');
  while (pos != string::npos) {
    inputArray.push_back(input_json_values.substr(0, pos));
    input_json_values = input_json_values.substr(pos + 1);
    pos = input_json_values.find('\n');
  }
  inputArray.push_back(input_json_values);

  Solution solution;
  string word = json::parse(inputArray.at(0));
  int k = json::parse(inputArray.at(1));
  return solution.minimumDeletions(word, k);
}

File: problems/problems_3085/solution.py
from bisect import bisect_left

from math import inf

import solution
from collections import Counter
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.minimumDeletions(*test_input)

    def minimumDeletions(self, word: str, k: int) -> int:
        counter = Counter(word)
        values = sorted(counter.values())
        n = len(values)
        ans = len(word)
        for i, v in enumerate(values):
            if i > 0 and values[i - 1] == v:
                continue
            cur, min_d, max_d = 0, v, v + k
            for j in range(i):
                if values[j] >= min_d:
                    break
                cur += values[j]
            for j in range(n - 1, i, -1):
                if values[j] <= max_d:
                    break
                cur += values[j] - max_d
            ans = min(ans, cur)
        return ans
